cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java

```diff
@@ -0,0 +1,263 @@
+   * false otherwise.
+   */
+  @Override
+  protected boolean selfSatisfied(ClusterModel clusterModel, BalancingProposal proposal) {
+    if (proposal.balancingAction() != BalancingAction.REPLICA_MOVEMENT) {
+      throw new IllegalStateException("Found balancing action " + proposal.balancingAction() +
+          " but expected replica movement.");
+    }
+    Broker destinationBroker = clusterModel.broker(proposal.destinationBrokerId());
+    if (destinationBroker.replicas().size() >= _balancingConstraint.maxReplicasPerBroker()) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Update goal state after one round of self-healing / rebalance.
+   *
+   *  @param clusterModel The state of the cluster.
+   * @param excludedTopics The topics that should be excluded from the optimization proposal.
+   */
+  @Override
+  protected void updateGoalState(ClusterModel clusterModel, Set<String> excludedTopics)
+      throws AnalysisInputException, OptimizationFailureException {
+    // One pass is sufficient to satisfy or alert impossibility of this goal.
+    // Sanity check to confirm that the final distribution has less than the allowed number of replicas per broker.
+    ensureReplicaCapacitySatisfied(clusterModel);
+    // Sanity check: No self-healing eligible replica should remain at a decommissioned broker.
+    AnalyzerUtils.ensureNoReplicaOnDeadBrokers(clusterModel);
+    finish();
+  }
+
+  /**
+   * Sanity check: Ensure that the replica capacity per broker is not exceeded.
+   *
+   * @param clusterModel The cluster model.
+   */
+  private void ensureReplicaCapacitySatisfied(ClusterModel clusterModel) {
+    for (Broker broker : brokersToBalance(clusterModel)) {
+      int numBrokerReplicas = broker.replicas().size();
+      if (numBrokerReplicas > _balancingConstraint.maxReplicasPerBroker()) {
+        throw new IllegalStateException(String.format("Replicas in broker: %d exceeds the maximum allowed number of "
+            + "replicas: %d.", numBrokerReplicas, _balancingConstraint.maxReplicasPerBroker()));
+      }
+    }
+  }
+
+  /**
+   * Rebalance the given broker without violating the constraints of the current goal and optimized goals.
+   *  @param broker         Broker to be balanced.
+   * @param clusterModel   The state of the cluster.
+
+   * @param optimizedGoals Optimized goals.
+
+   * @param excludedTopics The topics that should be excluded from the optimization proposals.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker, ClusterModel clusterModel, Set<Goal> optimizedGoals,
+      Set<String> excludedTopics) throws AnalysisInputException, ModelInputException {
+    LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+    int numBrokerReplicas = broker.replicas().size();
+    for (Replica replica : new ArrayList<>(broker.replicas())) {
+      if ((broker.isAlive() && numBrokerReplicas <= _balancingConstraint.maxReplicasPerBroker())
```

Comment:
Should this be checked before iterating over the replicas in the broker?

Reply:
Split the check into the cases for `break` and `continue`. (1) breaks when the alive broker is balanced. (2) continues when the topic is excluded but the first constraint is not satisfied.